rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+# SlateDB Checkpoints and Clones
+
+Status: Under Review
+
+Authors:
+* [Rohan Desai](https://github.com/rodesai)
+
+## Current Implementation
+
+SlateDB runs a Writer, a Compactor - optionally in a separate process, and a Garbage Collector (GC) - also optionally in a separate process. The writer writes data to L0 SSTs. The Compactor then goes and compacts these L0 SSTs into Sorted Runs (SR) under the “compacted” directory, and then further compacts those SRs as they accumulate. Updates to the current set of L0s and Sorted Runs are synchronized between the Writer and Compactor through updates to the centralized manifest. When reading, the Writer gets a reference to a copy of the metadata for the current set of sorted runs and searches them for the requested data. Meanwhile, the GC scans the SSTs in Object Storage and deletes any SSTs that are not referenced by the current manifest and are older than some threshold.
+
+## Problem
+
+The current implementation poses a few problems and limitations:
+1. There is nothing preventing the GC from deleting SSTs out from the Writer while it’s attempting to read their data. This isn’t a big deal today as we only support key lookups, which are fast. But this could become a more acute problem when we support range scans which can be long.
+1. Currently only the Writer process can interact with the database. We’d like to support additional read-only clients, called Readers that can read data.
+1. Some use cases can benefit from forking the database and establishing a new db that relies on the original DB’s SSTs but writes new data to a different set of SSTs from the original writer.
+
+## Goals
+
+In this RFC, we propose adding checkpoints and clones to address this limitation.
+- Clients will be able to establish cheap db checkpoints that reference the current version of SlateDB’s manifest.
+- Writers will use checkpoints to “pin” a version of the database for reads to avoid the GC from deleting the SSTs.
```

Comment:
what would be the intention for writers to pin a version?

Reply:
The intent was to avoid the GC deleting the SSTs used to serve reads. The GC section of the doc details this - the RFC proposes that the GC will be extended to look for SSTs that are referenced by a checkpoint and to not delete those SSTs. By maintaining a checkpoint we prevent the SSTs in the writer's current view of the db from being deleted.